### PR TITLE
Generate simple stubs code

### PR DESCRIPTION
**Pull request by gnossen**
This PR completes the implementation of simple stubs, which allows users to invoke RPCs without explicitly instantiating a channel or a stub. For the moment, this is an experimental API.

This enables the following code transformation:

```python
import grpc
import request_pb2
import test_service_pb2_grpc

with grpc.insecure_channel("localhost:50051") as channel:
  test_service_stub = test_service_pb2_grpc.TestServiceStub(channel)
  response = test_service_stub.UnaryCall(request_pb2.SimpleRequest)
```
to

```python
from request_pb2 import SimpleRequest
from test_service_pb2_grpc import TestService

response = TestService.UnaryCall(SimpleRequest(), "localhost:50051")
```

Example generated code:
```python
  # This class is part of an EXPERIMENTAL API.
class TestService(object):
    """Missing associated documentation comment in .proto file"""

    @staticmethod
    def UnaryCall(request,
            target,
            options=(),
            channel_credentials=None,
            call_credentials=None,
            compression=None,
            wait_for_ready=None,
            timeout=None,
            metadata=None):
        return grpc.experimental.unary_unary(request, target, '/grpc_protoc_plugin.TestService/UnaryCall',
            tests_dot_protoc__plugin_dot_protos_dot_requests_dot_r_dot_test__requests__pb2.SimpleRequest.SerializeToString,
            tests_dot_protoc__plugin_dot_protos_dot_responses_dot_test__responses__pb2.SimpleResponse.FromString,
            options, channel_credentials,
            call_credentials, compression, wait_for_ready, timeout, metadata)

    @staticmethod
    def StreamingOutputCall(request,
            target,
            options=(),
            channel_credentials=None,
            call_credentials=None,
            compression=None,
            wait_for_ready=None,
            timeout=None,
            metadata=None):
        return grpc.experimental.unary_stream(request, target, '/grpc_protoc_plugin.TestService/StreamingOutputCall',
            tests_dot_protoc__plugin_dot_protos_dot_requests_dot_r_dot_test__requests__pb2.StreamingOutputCallRequest.SerializeToString,
            tests_dot_protoc__plugin_dot_protos_dot_responses_dot_test__responses__pb2.StreamingOutputCallResponse.FromString,
            options, channel_credentials,
            call_credentials, compression, wait_for_ready, timeout, metadata)

    @staticmethod
    def StreamingInputCall(request_iterator,
            target,
            options=(),
            channel_credentials=None,
            call_credentials=None,
            compression=None,
            wait_for_ready=None,
            timeout=None,
            metadata=None):
        return grpc.experimental.stream_unary(request_iterator, target, '/grpc_protoc_plugin.TestService/StreamingInputCall',
            tests_dot_protoc__plugin_dot_protos_dot_requests_dot_r_dot_test__requests__pb2.StreamingInputCallRequest.SerializeToString,
            tests_dot_protoc__plugin_dot_protos_dot_responses_dot_test__responses__pb2.StreamingInputCallResponse.FromString,
            options, channel_credentials,
            call_credentials, compression, wait_for_ready, timeout, metadata)

    @staticmethod
    def FullDuplexCall(request_iterator,
            target,
            options=(),
            channel_credentials=None,
            call_credentials=None,
            compression=None,
            wait_for_ready=None,
            timeout=None,
            metadata=None):
        return grpc.experimental.stream_stream(request_iterator, target, '/grpc_protoc_plugin.TestService/FullDuplexCall',
            tests_dot_protoc__plugin_dot_protos_dot_requests_dot_r_dot_test__requests__pb2.StreamingOutputCallRequest.SerializeToString,
            tests_dot_protoc__plugin_dot_protos_dot_responses_dot_test__responses__pb2.StreamingOutputCallResponse.FromString,
            options, channel_credentials,
            call_credentials, compression, wait_for_ready, timeout, metadata)

    @staticmethod
    def HalfDuplexCall(request_iterator,
            target,
            options=(),
            channel_credentials=None,
            call_credentials=None,
            compression=None,
            wait_for_ready=None,
            timeout=None,
            metadata=None):
        return grpc.experimental.stream_stream(request_iterator, target, '/grpc_protoc_plugin.TestService/HalfDuplexCall',
            tests_dot_protoc__plugin_dot_protos_dot_requests_dot_r_dot_test__requests__pb2.StreamingOutputCallRequest.SerializeToString,
            tests_dot_protoc__plugin_dot_protos_dot_responses_dot_test__responses__pb2.StreamingOutputCallResponse.FromString,
            options, channel_credentials,
            call_credentials, compression, wait_for_ready, timeout, metadata)
```

### CODE REVIEW

File: src/python/grpcio_tests/commands.py [hunk 1]
```diff
@@ -193,6 +193,7 @@ class TestGevent(setuptools.Command):
         'unit._server_ssl_cert_config_test',
         # TODO(https://github.com/grpc/grpc/issues/14901) enable this test
         'protoc_plugin._python_plugin_test.PythonPluginTest',
+        'protoc_plugin._python_plugin_test.SimpleStubsPluginTest',
```
Comment (lidizheng) [inline src/python/grpcio_tests/commands.py]: Why is it banned for `gevent`?

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py [hunk 2]
```diff
@@ -503,5 +504,118 @@ def half_duplex_request_iterator():
         service.server.stop(None)
 
 
+@unittest.skipIf(sys.version_info[0] < 3, "Unsupported on Python 2.")
+class SimpleStubsPluginTest(unittest.TestCase):
+    servicer_methods = _ServicerMethods()
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: nit: Is `setUpClass` an option?

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py [hunk 3]
```diff
@@ -503,5 +504,118 @@ def half_duplex_request_iterator():
...
+        super(SimpleStubsPluginTest, self).tearDown()
+
+    def testUnaryCall(self):
+        request = request_pb2.SimpleRequest(response_size=13)
+        response = service_pb2_grpc.TestService.UnaryCall(
+            request,
+            self._target,
+            channel_credentials=grpc.experimental.insecure_channel_credentials(
+            ),
+            wait_for_ready=True)
```
Comment (lidizheng) [inline src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: nit: Can we simplify the `grpc.experimental.insecure_channel_credentials` design?
nit: Is `wait_for_ready=True` necessary for tests to pass on Windows? If the underlying channel is pooled, can we warm up the channel object earlier to prevent explicitly setting this flag?

File: src/compiler/python_generator.cc [hunk 4]
```diff
@@ -570,6 +570,92 @@ bool PrivateGenerator::PrintAddServicerToServer(
...
+    const grpc_generator::Service* service, grpc_generator::Printer* out) {
+  StringMap dict;
+  dict["Service"] = service->name();
+  out->Print("\n\n");
+  out->Print(" # This class is part of an EXPERIMENTAL API.\n");
+  out->Print(dict, "class $Service$(object):\n");
+  {
+    IndentScope class_indent(out);
+    StringVector service_comments = service->GetAllComments();
+    PrintAllComments(service_comments, out);
```
Comment (lidizheng) [inline src/compiler/python_generator.cc]: Maybe we can skip the `pass` in your example generated code somewhere here?

File: src/compiler/python_private_generator.h [hunk 5]
```diff
@@ -59,6 +59,9 @@ struct PrivateGenerator {
                  const grpc_generator::Service* service,
                  grpc_generator::Printer* out);
 
+  bool PrintServiceClass(const grpc::string& package_qualified_service_name,
+                         const grpc_generator::Service* service,
+                         grpc_generator::Printer* out);
```
Comment (lidizheng) [inline src/compiler/python_private_generator.h]: optional: Should we try to make the generated code PEP8-complied? https://github.com/grpc/grpc/issues/5014
nit: Can we check if linters (`yapf`, `pylint`) are happy with the generated code?

File: src/compiler/python_generator.cc [hunk 6]
```diff
@@ -570,6 +570,92 @@ bool PrivateGenerator::PrintAddServicerToServer(
...
 }
 
+/* Prints out a service class used as a container for static methods pertaining
+ * to a class. This class has the exact name of service written in the ".proto"
+ * file, with no suffixes. Since this class merely acts as a namespace, it
+ * should never be instantiated.
+ */
+bool PrivateGenerator::PrintServiceClass(
+    const grpc::string& package_qualified_service_name,
+    const grpc_generator::Service* service, grpc_generator::Printer* out) {
```
Comment (lidizheng) [inline src/compiler/python_generator.cc]: Should we call it service class? It is a stub, but is named as service? Do you think something like `SimpleStubClass` is a viable name?

File: src/python/grpcio_tests/commands.py (hunk 1 above)
Comment (gnossen) [reply src/python/grpcio_tests/commands.py]: [The other test in this file also doesn't work.](https://github.com/grpc/grpc/blob/d8623ba548cc9d65d98bf136d1bd52b7aba1c219/src/python/grpcio_tests/commands.py#L195) After seeing that the failure was due to the setup in the rest of the file, I didn't debug it further.

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py (hunk 2 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: Yes. But it's not the pattern that the rest of the tests in this file are using. I decided to prioritize meshing well with the rest of the file over going with something slightly more conventional.

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py (hunk 3 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: That's rather vague. How do you mean? Were you expecting insecure to be default? Were you expecting the name to be shorter? FWIW, the default is currently actually `LocalCredentials`, which I was originally using here. But I ran into  https://github.com/grpc/grpc/issues/22257, which is why I'm explicitly specifying here. Eric and Sanjay both have weighed in that explicit is better than implicit for security reasons.

`wait_for_ready` is necessary in general to make this test not flaky. Otherwise, there's a race condition between the server thread coming ready and the main test thread sending the request. I don't think relying on the implementation detail that channels are pooled is a good idea. `wait_for_ready` is already a standard solution for this. However, we *could* flip the default to `True` for `wait_for_ready`, but we should probably have a bit more of a conversation about that first.

File: src/compiler/python_generator.cc (hunk 4 above)
Comment (gnossen) [reply src/compiler/python_generator.cc]: Thought I was following a pattern from the existing generated code, but it turns out to be something only in the beta generated code. I'll switch this over.

File: src/compiler/python_private_generator.h (hunk 5 above)
Comment (gnossen) [reply src/compiler/python_private_generator.h]: That seems reasonable except perhaps for snake case. I've seen people complain in some languages that the casing matches their languages' preferences and I've seen other complain that it doesn't match what they wrote in their ".proto" file. One point against this is that the `protobuf` module is *not* going to do this, so they'd end up mixing snake case with camel case, which IMO, is worse than the situation we currently have.

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py (hunk 3 above)
Comment (lidizheng) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: The `LocalCredentials` issue is not meant to be portable, maybe we should consider other default value here? Should we use `insecure_channel` by default? It seems least surprising to our users.

This issue seems more or less test-specific. We could try to use `channel_ready_future` in `setUp` to ensure the server is ready, or other mechanism. This can be in a separate PR.

About flipping the default value for `wait_for_ready`, in usual scenario, gRPC Python are connecting to some public services, where fail-fast helps users to debug their configuration error. So, I think it should stay `False`.

My intent is to make this test cleaner. Ideally, if we share this test case to users, they can immediately agree that the new API is much more simpler.

File: src/compiler/python_generator.cc (hunk 6 above)
Comment (gnossen) [reply src/compiler/python_generator.cc]: I'm not married to the name here since this class is just an implementation detail. My hope was that in the future, this class will also serve as a place to add things that pertain to the server. It's called "Service" here because it corresponds one-to-one to your Protobuf-level `service` definition. If you feel strongly, I'll change it.

Comment (lidizheng) [reply src/compiler/python_generator.cc]: I'm fine with the naming if we are going to add servicer definition someday. Should we create a TODO issue for it? Or point it to an existing design doc?

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py (hunk 3 above)
Comment (gnossen) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: I definitely agree that `LocalCredentials` cannot remain the default.

Everyone I have talked to about it agrees that is not acceptable for an API to be insecure by default. So `insecure_channel` is out as well.

You could use `ssl_credentials` as your default option. But that doesn't make sense inside of google3. You'd want `loas2` there instead.

You could do something more complicated, like the `CompositeCredentials` idea I proposed a while ago. But, even in a tiered solution like this, you'd need to pick between SSL and LOAS2.

In short, this is a complicated issue. It's one that I think deserves more attention, and will probably require convincing other parts of the team. For the moment, I'd like to remove the default credential, since it is always backwards compatible to add one in.

**Comment by gnossen** [comment]
Regarding an example: My hope was to change the `helloworld` client code once this API graduates from experimental. But you make a good point. I'll create a separate example for the short term clearly indicating that this is currently an experimental API.

File: src/compiler/python_generator.cc (hunk 4 above)
Comment (lidizheng) [reply src/compiler/python_generator.cc]: Is this change possible?

File: src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py (hunk 3 above)
Comment (lidizheng) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: > For the moment, I'd like to remove the default credential, since it is always backwards compatible to add one in.

Changing default value can be surprising to users. This might be in the scope of PR, but we might need a better solution.

Comment (gnossen) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: This is an unreleased experimental API that emits a warning when you use it. Removing the default value is definitely fair game. As I stated above, I would like to look into a default value for this feature, but let's not tie it to this PR.

Comment (gnossen) [reply src/python/grpcio_tests/tests/protoc_plugin/_python_plugin_test.py]: Temporarily removed the default and added a TODO. I'm looking into some longer term solutions to this problem.

File: src/compiler/python_generator.cc (hunk 6 above)
Comment (gnossen) [reply src/compiler/python_generator.cc]: TODO added.

File: src/compiler/python_private_generator.h (hunk 5 above)
Comment (gnossen) [reply src/compiler/python_private_generator.h]: Updated to 4 spaces.

File: src/compiler/python_generator.cc (hunk 4 above)
Comment (gnossen) [reply src/compiler/python_generator.cc]: Fixed for not just this class, but all generated code.

**Comment by gnossen** [comment]
Failures:
 - https://github.com/grpc/grpc/issues/22281
 - https://github.com/grpc/grpc/issues/22282

**Comment by TomaszBorczyk** [comment]
Since this is experimental API, it would be nice to have a flag that enables/disables generation of this part of code.
Also, this feature is largely undocumented: this issue is the only source I've found that explains what it does, and it seems this code is generated automatically every time
